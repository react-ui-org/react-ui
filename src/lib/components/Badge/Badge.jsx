import PropTypes from 'prop-types';
import React from 'react';
import { withProviderContext } from '../../provider';
import { classNames } from '../../utils/classNames';
import { getRootColorClassName } from '../_helpers/getRootColorClassName';
import styles from './Badge.scss';

export const Badge = ({
  color,
  id,
  label,
  priority,
}) => (
  <div
    className={classNames(
      styles.root,
      priority === 'outline' && styles.rootPriorityOutline,
      getRootColorClassName(color, styles),
    )}
    id={id}
  >
    {label}
  </div>
);

Badge.defaultProps = {
  color: 'note',
  id: undefined,
  priority: 'filled',
};

Badge.propTypes = {
  /**
   * [Color variant](/foundation/colors#component-colors) to clarify importance and meaning of the badge.
   */
  color: PropTypes.oneOf(
    ['primary', 'secondary', 'success', 'warning', 'danger', 'help', 'info', 'note', 'light', 'dark'],
  ),
  /**
   * ID of the root HTML element.
   */
  id: PropTypes.string,
  /**
   * Text to be displayed.
   */
  label: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]).isRequired,
  /**
   * Visual priority to highlight or suppress the badge.
   */
  priority: PropTypes.oneOf(['filled', 'outline']),
};

export const BadgeWithContext = withProviderContext(Badge, 'Badge');

export default BadgeWithContext;
