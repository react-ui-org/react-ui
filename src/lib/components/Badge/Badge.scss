@use "sass:map";
@use "sass:math";
@use "../../styles/theme/borders";
@use "../../styles/theme/colors";
@use "../../styles/theme/typography";

$_badge-size: 1.25rem;

.root {
    display: inline-block;
    min-width: $_badge-size;
    height: $_badge-size;
    padding: 0.25rem 0.35rem;
    overflow: hidden;
    font-weight: map.get(typography.$font-weight-values, bold);
    font-size: typography.$size-smaller;
    line-height: 1;
    text-align: center;
    white-space: nowrap;
    vertical-align: baseline;
    color: var(--rui-local-color);
    border-radius: math.div($_badge-size, 2);
    background-color: var(--rui-local-background-color);
    box-shadow: var(--rui-local-box-shadow, #{0 0 0 2px rgb(255 255 255 / 80%)});
}

.rootColorPrimary {
    --rui-local-color: #{map.get(colors.$brand, on-primary)};
    --rui-local-background-color: #{map.get(colors.$brand, primary)};
}

.rootColorSecondary {
    --rui-local-color: #{map.get(colors.$brand, on-secondary)};
    --rui-local-background-color: #{map.get(colors.$brand, secondary)};
}

.rootColorSuccess {
    --rui-local-color: #{map.get(colors.$ui, on-success)};
    --rui-local-background-color: #{map.get(colors.$ui, success)};
}

.rootColorWarning {
    --rui-local-color: #{map.get(colors.$ui, on-warning)};
    --rui-local-background-color: #{map.get(colors.$ui, warning)};
}

.rootColorDanger {
    --rui-local-color: #{map.get(colors.$ui, on-danger)};
    --rui-local-background-color: #{map.get(colors.$ui, danger)};
}

.rootColorHelp {
    --rui-local-color: #{map.get(colors.$ui, on-help)};
    --rui-local-background-color: #{map.get(colors.$ui, help)};
}

.rootColorInfo {
    --rui-local-color: #{map.get(colors.$ui, on-info)};
    --rui-local-background-color: #{map.get(colors.$ui, info)};
}

.rootColorNote {
    --rui-local-color: #{map.get(colors.$ui, on-note)};
    --rui-local-background-color: #{map.get(colors.$ui, note)};
}

.rootColorLight {
    --rui-local-color: #{colors.$dark};
    --rui-local-background-color: #{colors.$light};
    --rui-local-box-shadow: none;
}

.rootColorDark {
    --rui-local-color: #{colors.$light};
    --rui-local-background-color: #{colors.$dark};
    --rui-local-box-shadow: none;
}

.rootPriorityOutline {
    --rui-local-background-color: transparent;
    --rui-local-box-shadow: none;

    padding-top: 0.1875rem;
    padding-bottom: 0.1875rem;
    border: borders.$width solid currentColor;
}

.rootPriorityOutline.rootColorPrimary {
    --rui-local-color: #{map.get(colors.$brand, primary)};
}

.rootPriorityOutline.rootColorSecondary {
    --rui-local-color: #{map.get(colors.$brand, secondary)};
}

.rootPriorityOutline.rootColorSuccess {
    --rui-local-color: #{map.get(colors.$ui, success)};
}

.rootPriorityOutline.rootColorWarning {
    --rui-local-color: #{map.get(colors.$ui, warning)};
}

.rootPriorityOutline.rootColorDanger {
    --rui-local-color: #{map.get(colors.$ui, danger)};
}

.rootPriorityOutline.rootColorHelp {
    --rui-local-color: #{map.get(colors.$ui, help)};
}

.rootPriorityOutline.rootColorInfo {
    --rui-local-color: #{map.get(colors.$ui, info)};
}

.rootPriorityOutline.rootColorNote {
    --rui-local-color: #{map.get(colors.$ui, note)};
}

.rootPriorityOutline.rootColorLight {
    --rui-local-color: #{colors.$light};
}

.rootPriorityOutline.rootColorDark {
    --rui-local-color: #{colors.$dark};
}
