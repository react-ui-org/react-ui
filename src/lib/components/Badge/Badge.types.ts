export interface BadgeProps {
  /**
   * Rest of the props.
   */
  [key: string]: unknown;
  /**
   * [Color variant](/foundation/colors#component-colors) to clarify importance and meaning of the badge.
   */
  color?: Color;
  /**
   * Text to be displayed.
   */
  label: string;
  /**
   * Visual priority to highlight or suppress the badge.
   */
  priority?: Exclude<Priority, 'flat'>;
}
