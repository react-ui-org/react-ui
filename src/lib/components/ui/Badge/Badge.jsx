import PropTypes from 'prop-types';
import React from 'react';
import { withProviderContext } from '../../../provider';
import styles from './Badge.scss';

export const Badge = (props) => {
  const {
    id,
    label,
    priority,
    type,
  } = props;

  let rootTypeClass = styles.note;
  let rootPriorityClass = '';

  if (type) {
    if (type === 'success') {
      rootTypeClass = styles.success;
    } else if (type === 'error') {
      rootTypeClass = styles.error;
    } else if (type === 'warning') {
      rootTypeClass = styles.warning;
    } else if (type === 'info') {
      rootTypeClass = styles.info;
    } else if (type === 'help') {
      rootTypeClass = styles.help;
    } else if (type === 'light') {
      rootTypeClass = styles.light;
    }
  }

  if (priority === 'outline') {
    rootPriorityClass = styles.outline;
  }

  return (
    <div
      className={(`
        ${styles.root}
        ${rootPriorityClass}
        ${rootTypeClass}
      `).trim()}
      id={id}
    >
      {label}
    </div>
  );
};

Badge.defaultProps = {
  id: undefined,
  priority: 'filled',
  type: 'note',
};

Badge.propTypes = {
  /**
   * ID of the root HTML element.
   */
  id: PropTypes.string,
  /**
   * Text to be displayed.
   */
  label: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]).isRequired,
  /**
   * Visual priority to highlight or suppress the badge.
   */
  priority: PropTypes.oneOf(['filled', 'outline']),
  /**
   * Color variant to clarify importance and meaning of the badge.
   */
  type: PropTypes.oneOf(['error', 'help', 'info', 'note', 'success', 'warning', 'light']),
};

export const BadgeWithContext = withProviderContext(Badge, 'Badge');

export default BadgeWithContext;
