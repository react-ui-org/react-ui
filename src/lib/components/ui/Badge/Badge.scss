@use '../../../styles/settings/borders';
@use '../../../styles/settings/colors';
@use '../../../styles/settings/typography';

$_badge-size: 1.25rem;

.root {
  display: inline-block;
  min-width: $_badge-size;
  height: $_badge-size;
  padding: 0.25rem 0.35rem;
  overflow: hidden;
  font-weight: map-get(typography.$font-weight-values, bold);
  font-size: typography.$size-smaller;
  line-height: 1;
  text-align: center;
  white-space: nowrap;
  vertical-align: baseline;
  border-radius: $_badge-size / 2;
  box-shadow: 0 0 0 2px rgba(255, 255, 255, 0.8);
}

.success {
  color: map-get(colors.$ui, on-success);
  background-color: map-get(colors.$ui, success);
}

.warning {
  color: map-get(colors.$ui, on-warning);
  background-color: map-get(colors.$ui, warning);
}

.error {
  color: map-get(colors.$ui, on-error);
  background-color: map-get(colors.$ui, error);
}

.info {
  color: map-get(colors.$ui, on-info);
  background-color: map-get(colors.$ui, info);
}

.help {
  color: map-get(colors.$ui, on-help);
  background-color: map-get(colors.$ui, help);
}

.note {
  color: map-get(colors.$ui, on-note);
  background-color: map-get(colors.$ui, note);
}

.light {
  color: colors.$black;
  background-color: colors.$white;
  box-shadow: none;
}

.outline {
  padding-top: 0.1875rem;
  padding-bottom: 0.1875rem;
  border: borders.$width solid currentColor;
  background: none;
  box-shadow: none;
}

.outline.success {
  color: map-get(colors.$ui, success);
}

.outline.warning {
  color: map-get(colors.$ui, warning);
}

.outline.error {
  color: map-get(colors.$ui, error);
}

.outline.info {
  color: map-get(colors.$ui, info);
}

.outline.help {
  color: map-get(colors.$ui, help);
}

.outline.note {
  color: map-get(colors.$ui, note);
}

.outline.light {
  color: colors.$white;
}
