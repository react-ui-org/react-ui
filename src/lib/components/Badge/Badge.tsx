import React from 'react';
import { withGlobalProps } from '../../provider';
import { transferProps } from '../_helpers/transferProps';
import { classNames } from '../../utils/classNames';
import { getRootColorClassName } from '../_helpers/getRootColorClassName';
import styles from './Badge.scss';
import { BadgeProps } from './Badge.types';

export const Badge: React.FunctionComponent<BadgeProps> = ({
  color = 'note',
  label,
  priority = 'filled',
  ...restProps
}) => (
  <div
    {...transferProps(restProps)}
    className={classNames(
      styles.root,
      priority === 'outline' && styles.isRootPriorityOutline,
      getRootColorClassName(color, styles),
    )}
  >
    {label}
  </div>
);

export const BadgeWithGlobalProps = withGlobalProps(Badge, 'Badge');

export default BadgeWithGlobalProps;
