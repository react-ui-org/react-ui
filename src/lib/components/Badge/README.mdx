---
name: Badge
menu: 'Miscellaneous'
route: /components/badge
---

# Badge

Badge highlights content with short additional information.

import {
  Playground,
  Props,
} from 'docz'
import Placeholder from '../../../docs/_components/Placeholder'
import { Badge } from './Badge'

## Basic Usage

To implement the Badge component, you need to import it first:

```js
import { Badge } from '@react-ui-org/react-ui';
```

And use it:

<Playground>
  <Badge label="Hello!" />
</Playground>

See [API](#api) for all available options.

## General Guidelines

Use badges to highlight content with a **short and brief** additional
information: a number, a word, but no more than two words.

## Priorities

There are two **visual priorities** of badges to choose from, from highest to
lowest:

1. filled
2. outline

All priorities come in supported
[component colors](/foundation/colors#component-colors).

### Filled

High-emphasis priority to draw user's attention.

<Playground>
  <Badge label="1" color="primary" />
  <Badge label="2" color="secondary" />
  <Badge label="3" color="success" />
  <Badge label="4" color="warning" />
  <Badge label="5" color="danger" />
  <Badge label="10" color="help" />
  <Badge label="34" color="info" />
  <Badge label="99+" />
  <Placeholder dark inline>
    <Badge label="365" color="light" />
  </Placeholder>
  <Badge label="999+" color="dark" />
</Playground>

<Playground>
  <Badge label="Primary" color="primary" />
  <Badge label="Secondary" color="secondary" />
  <Badge label="Success" color="success" />
  <Badge label="Warning" color="warning" />
  <Badge label="Danger" color="danger" />
  <Badge label="Help" color="help" />
  <Badge label="Info" color="info" />
  <Badge label="Note" />
  <Placeholder dark inline>
    <Badge label="Light" color="light" />
  </Placeholder>
  <Badge label="Dark" color="dark" />
</Playground>

### Outline

Medium-emphasis priority to provide additional context in an unobtrusive way.

<Playground>
  <Badge priority="outline" label="1" color="primary" />
  <Badge priority="outline" label="2" color="secondary" />
  <Badge priority="outline" label="3" color="success" />
  <Badge priority="outline" label="4" color="warning" />
  <Badge priority="outline" label="5" color="danger" />
  <Badge priority="outline" label="10" color="help" />
  <Badge priority="outline" label="34" color="info" />
  <Badge priority="outline" label="99+" />
  <Placeholder dark inline>
    <Badge priority="outline" label="365" color="light" />
  </Placeholder>
  <Badge priority="outline" label="999+" color="dark" />
</Playground>

<Playground>
  <Badge priority="outline" label="Primary" color="primary" />
  <Badge priority="outline" label="Secondary" color="secondary" />
  <Badge priority="outline" label="Success" color="success" />
  <Badge priority="outline" label="Warning" color="warning" />
  <Badge priority="outline" label="Danger" color="danger" />
  <Badge priority="outline" label="Help" color="help" />
  <Badge priority="outline" label="Info" color="info" />
  <Badge priority="outline" label="Note" />
  <Placeholder dark inline>
    <Badge priority="outline" label="Light" color="light" />
  </Placeholder>
  <Badge priority="outline" label="Dark" color="dark" />
</Playground>

## API

<Props table of={Badge} />
