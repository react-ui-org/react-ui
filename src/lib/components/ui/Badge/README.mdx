---
name: Badge
menu: 'UI'
route: /components/ui/badge
---

# Badge

Badge highlights content with short additional information.

import {
  Playground,
  Props,
} from 'docz'
import { Badge } from './Badge'

## Basic Usage

To implement the Badge component, you need to import it first:

```js
import { Badge } from '@react-ui-org/react-ui';
```

And use it:

<Playground>
  <Badge label="Hello!" />
</Playground>

See [API](#api) for all available options.

## General Guidelines

Use badges to highlight content with a **short and brief** additional
information: a number, a word, but no more than two words.

## Priorities

There are two **visual priorities** of badges to choose from, from highest to
lowest:

1. filled
2. outline

All priorities come in following **color variants** (types):

* all [UI colors](/foundation/colors#ui-colors): success, warning, error, help,
  info, note
* light color

### Filled

High-emphasis priority to draw user's attention.

<Playground>
  <Badge label="1" type="success" />
  <Badge label="2" type="warning" />
  <Badge label="3" type="error" />
  <Badge label="10" type="help" />
  <Badge label="34" type="info" />
  <Badge label="99+" type="note" />
  <Badge label="365" type="light" />
</Playground>

<Playground>
  <Badge label="Success" type="success" />
  <Badge label="Warning" type="warning" />
  <Badge label="Error" type="error" />
  <Badge label="Help" type="help" />
  <Badge label="Info" type="info" />
  <Badge label="Note" type="note" />
  <Badge label="Light" type="light" />
</Playground>

### Outline

Medium-emphasis priority to provide additional context in an unobtrusive way.

<Playground>
  <Badge label="1" priority="outline" type="success" />
  <Badge label="2" priority="outline" type="warning" />
  <Badge label="3" priority="outline" type="error" />
  <Badge label="10" priority="outline" type="help" />
  <Badge label="34" priority="outline" type="info" />
  <Badge label="99+" priority="outline" type="note" />
  <Badge label="365" priority="outline" type="light" />
</Playground>

<Playground>
  <Badge label="Success" priority="outline" type="success" />
  <Badge label="Warning" priority="outline" type="warning" />
  <Badge label="Error" priority="outline" type="error" />
  <Badge label="Help" priority="outline" type="help" />
  <Badge label="Info" priority="outline" type="info" />
  <Badge label="Note" priority="outline" type="note" />
  <Badge label="Light" priority="outline" type="light" />
</Playground>

## API

<Props table of={Badge} />
