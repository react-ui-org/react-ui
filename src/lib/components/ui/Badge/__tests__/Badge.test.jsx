import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import { Badge } from '../Badge';

describe('rendering', () => {
  it('renders correctly with number label', () => {
    const tree = shallow(
      <Badge label={3} />,
    );

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly with string label', () => {
    const tree = shallow(
      <Badge label="label" />,
    );

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly with all props', () => {
    const tree = shallow(
      <Badge
        id="custom-id"
        label={3}
        priority="outline"
        type="warning"
      />,
    );

    expect(shallowToJson(tree)).toMatchSnapshot();
  });
});
