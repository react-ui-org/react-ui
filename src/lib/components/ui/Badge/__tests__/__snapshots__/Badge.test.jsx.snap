// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly with all props 1`] = `
<div
  className="root
        outline
        warning"
  id="custom-id"
>
  3
</div>
`;

exports[`rendering renders correctly with number label 1`] = `
<div
  className="root
        
        note"
>
  3
</div>
`;

exports[`rendering renders correctly with string label 1`] = `
<div
  className="root
        
        note"
>
  label
</div>
`;
